// resium/dist/resium.js
import { createContext as e, forwardRef as t, useCallback as n, useContext as r, useEffect as i, useImperativeHandle as a, useLayoutEffect as o, useMemo as s, useReducer as c, useRef as l, useState as u } from "react";
import { BillboardCollection as d, BillboardGraphics as f, BoxGraphics as p, BufferPoint as m, BufferPointCollection as h, BufferPolygon as g, BufferPolygonCollection as _, BufferPolyline as v, BufferPolylineCollection as ee, Cesium3DTilesTerrainProvider as y, Cesium3DTileset as b, Cesium3DTilesetGraphics as x, CesiumWidget as S, ClassificationPrimitive as C, CloudCollection as w, Color as T, CorridorGraphics as te, CubeMapPanorama as E, CustomDataSource as ne, CylinderGraphics as D, CzmlDataSource as O, EllipseGraphics as k, EllipsoidGraphics as A, Entity as j, EquirectangularPanorama as M, Event as re, Fog as N, GeoJsonDataSource as P, GeoJsonPrimitive as F, Google2DImageryProvider as ie, GroundPolylinePrimitive as ae, GroundPrimitive as oe, ImageryLayer as se, KmlDataSource as ce, LabelCollection as le, LabelGraphics as ue, MVTDataProvider as de, Material as fe, Model as pe, ModelGraphics as me, Moon as he, ParticleSystem as ge, PathGraphics as _e, PlaneGraphics as ve, PointGraphics as ye, PointPrimitiveCollection as be, PolygonGraphics as xe, PolylineCollection as Se, PolylineGraphics as Ce, PolylineVolumeGraphics as we, PostProcessStage as Te, PostProcessStageComposite as Ee, PostProcessStageLibrary as I, Primitive as De, RectangleGraphics as Oe, Resource as L, SceneMode as R, ScreenSpaceEventHandler as ke, ScreenSpaceEventType as z, Sun as Ae, TimeDynamicPointCloud as je, Viewer as Me, WallGraphics as Ne, WebMapTileServiceImageryProvider as Pe, createGooglePhotorealistic3DTileset as Fe } from "cesium";
import { Fragment as B, jsx as V } from "react/jsx-runtime";
import { createPortal as Ie } from "react-dom";
//#region src/core/context.ts
var H = e({}), { Provider: Le, Consumer: Re } = H, U = () => r(H) || {}, W = (e, t) => {
	let n = (e) => {
		let n = U(), r = l(void 0), a = l(!1);
		return i(() => () => {
			n.camera && e.cancelFlightOnUnmount && n.camera.cancelFlight();
		}, [n.camera, e.cancelFlightOnUnmount]), i(() => {
			n.camera && n.scene && !n.scene.isDestroyed() && (!e.once || !a.current) && (n.camera.cancelFlight(), t(n.camera, e, r.current), a.current = !0), r.current = e;
		}), null;
	};
	return n.displayName = e, n;
};
//#endregion
//#region src/core/util.ts
function ze(e, t) {
	return t ? G(e).reduce((e, [n, r]) => (K(t, n) && (e[n] = r), e), {}) : {};
}
function G(e) {
	return Object.keys(e).map((t) => [t, e[t]]);
}
function K(e, t) {
	return !!e && e.indexOf(t) !== -1;
}
function Be(e, t) {
	return !!e && !!t && [...Object.keys(e), ...Object.keys(t)].every((n) => e[n] === t[n]);
}
function Ve(e) {
	return e && typeof e.isDestroyed == "function" && typeof e.destroy == "function";
}
function He(e) {
	return Ve(e) && e.isDestroyed();
}
function q(e) {
	return e && typeof e == "object" && "then" in e && typeof e.then == "function";
}
//#endregion
//#region src/core/EventManager.ts
var J = "__RESIUM_EVENT_MANAGER", Y = [
	"onClick",
	"onDoubleClick",
	"onMouseDown",
	"onMouseUp",
	"onMiddleClick",
	"onMiddleDown",
	"onMiddleUp",
	"onMouseMove",
	"onPinchEnd",
	"onPinchMove",
	"onPinchStart",
	"onRightClick",
	"onRightDown",
	"onRightUp",
	"onWheel",
	"onMouseEnter",
	"onMouseLeave"
], Ue = class e {
	static eventTypeMap = {
		onClick: z.LEFT_CLICK,
		onDoubleClick: z.LEFT_DOUBLE_CLICK,
		onMouseDown: z.LEFT_DOWN,
		onMouseUp: z.LEFT_UP,
		onMiddleClick: z.MIDDLE_CLICK,
		onMiddleDown: z.MIDDLE_DOWN,
		onMiddleUp: z.MIDDLE_UP,
		onMouseMove: z.MOUSE_MOVE,
		onPinchEnd: z.PINCH_END,
		onPinchMove: z.PINCH_MOVE,
		onPinchStart: z.PINCH_START,
		onRightClick: z.RIGHT_CLICK,
		onRightDown: z.RIGHT_DOWN,
		onRightUp: z.RIGHT_UP,
		onWheel: z.WHEEL,
		onMouseEnter: z.MOUSE_MOVE,
		onMouseLeave: z.MOUSE_MOVE
	};
	scene;
	sshe;
	events = {
		onClick: /* @__PURE__ */ new Map(),
		onDoubleClick: /* @__PURE__ */ new Map(),
		onMouseDown: /* @__PURE__ */ new Map(),
		onMouseUp: /* @__PURE__ */ new Map(),
		onMiddleClick: /* @__PURE__ */ new Map(),
		onMiddleDown: /* @__PURE__ */ new Map(),
		onMiddleUp: /* @__PURE__ */ new Map(),
		onMouseMove: /* @__PURE__ */ new Map(),
		onPinchEnd: /* @__PURE__ */ new Map(),
		onPinchMove: /* @__PURE__ */ new Map(),
		onPinchStart: /* @__PURE__ */ new Map(),
		onRightClick: /* @__PURE__ */ new Map(),
		onRightDown: /* @__PURE__ */ new Map(),
		onRightUp: /* @__PURE__ */ new Map(),
		onWheel: /* @__PURE__ */ new Map(),
		onMouseEnter: /* @__PURE__ */ new Map(),
		onMouseLeave: /* @__PURE__ */ new Map()
	};
	hovered = void 0;
	constructor(e) {
		this.scene = e, this.sshe = new ke(e?.canvas);
	}
	destroy() {
		this.hovered = void 0, this.sshe.isDestroyed() || this.sshe.destroy();
	}
	isDestroyed() {
		return this.sshe.isDestroyed();
	}
	on(e, t, n) {
		e && t === "onWheel" || this.events[t].set(e, n);
	}
	off(e, t) {
		this.events[t].delete(e), this.hovered === e && (this.hovered = void 0);
	}
	setEvents(e, t) {
		G(t).forEach(([t, n]) => {
			let r = t;
			K(Y, r) && (n ? this.on(e, r, n) : this.off(e, r));
		}), this.commit();
	}
	clearEvents(e) {
		this.hovered = void 0, Y.forEach((t) => {
			this.off(e, t);
		}), this.commit();
	}
	commit() {
		let t = this.sshe, n = this.sshe.isDestroyed();
		n || (this.events.onMouseEnter.size === 0 && this.events.onMouseLeave.size === 0 && this.events.onMouseMove.size === 0 ? this.sshe.removeInputAction(z.MOUSE_MOVE) : this.sshe.getInputAction(z.MOUSE_MOVE) || this.sshe.setInputAction(this.onMouseMove, z.MOUSE_MOVE)), G(this.events).forEach(([r, i]) => {
			if (r === "onMouseEnter" || r === "onMouseLeave" || r === "onMouseMove") return;
			let a = e.eventTypeMap[r];
			n || (i.size === 0 ? t.removeInputAction(a) : t.getInputAction(a) || t.setInputAction(this.eventCallback(r), a));
		});
	}
	getScreenSpaceEventHandler() {
		return this.sshe;
	}
	getEventCallback(e, t) {
		return t === null ? this.events[e].get(null) : this.events[e].get(t.id) || this.events[e].get(t.id?.entityCollection?.owner) || this.events[e].get(t.primitive) || this.events[e].get(t.tileset);
	}
	onMouseMove = (e) => {
		let t = this.pick(e.endPosition);
		this.hovered !== t && (this.hovered && (this.getEventCallback("onMouseLeave", this.hovered)?.(e, this.hovered), this.getEventCallback("onMouseLeave", null)?.(e, this.hovered)), t && (this.getEventCallback("onMouseEnter", t)?.(e, t), this.getEventCallback("onMouseEnter", null)?.(e, t))), t && this.getEventCallback("onMouseMove", t)?.(e, t), this.getEventCallback("onMouseMove", null)?.(e, t), this.hovered = t;
	};
	eventCallback = (e) => (t) => {
		let n = this.pick(t?.position);
		n && this.getEventCallback(e, n)?.(t, n), this.getEventCallback(e, null)?.(t, n);
	};
	pick(e) {
		if (e) return this.scene?.pick(e);
	}
}, We = ({ name: e, create: t, destroy: r, provide: s, update: d, cesiumReadonlyProps: f, cesiumEventProps: p, otherProps: m, setCesiumPropsAfterCreate: h, useCommonEvent: g, useRootEvent: _ }, v, ee) => {
	let y = l(void 0), b = U(), x = l(s ? {} : void 0), S = l({}), C = l(Ge(v)), w = l({}), [T, te] = u(!1), E = l(!1), [, ne] = c((e) => e + 1, 0), D = l(null), O = l(void 0), k = b?.[J], A = l(void 0), j = l(void 0), M = n(async (t) => {
		if (!y.current) return;
		let n = y.current, r = Object.keys(t), i = Object.keys(p || []), a = r.concat(Object.keys(w.current).filter((e) => !r.includes(e))).filter((e) => w.current[e] !== t[e]).map((e) => [
			e,
			w.current[e],
			t[e]
		]), o = [];
		for (let [e, t, r] of a) if (f?.includes(e)) o.push(e);
		else if (K(i, e)) {
			let i = p?.[e], a = n[i];
			a instanceof re && (t === void 0 ? (a.addEventListener(r), S.current[i] = r) : r === void 0 ? (a.removeEventListener(t), delete S.current[i]) : (a.removeEventListener(t), a.addEventListener(r)));
		} else e !== "children" && !Y.includes(e) && !m?.includes(e) && (n[e] = r);
		let s = _ ? x.current?.[J] : k;
		if (g && s && y.current && s.setEvents(_ ? null : y.current, t), d && E.current) {
			let e = d(y.current, t, w.current, b);
			q(e) && await e;
		}
		w.current = t, C.current = t, E.current && o.length > 0 && (process.env.NODE_ENV !== "production" && console.warn(`Warning: <${e}> is recreated because following read-only props have been updated: ${o.join(", ")}`), P(), await F(), A.current = N());
	}, [b]), N = n(async () => {
		await new Promise((e) => queueMicrotask(() => e(void 0)));
		let e = t?.(b, C.current, D.current), n;
		if (n = q(e) ? await e : e, Array.isArray(n) ? (y.current = n[0], O.current = n[1]) : y.current = n, h) await M(C.current);
		else {
			if (y.current && p) {
				let e = y.current;
				for (let t of Object.keys(C.current)) {
					let n = p[t];
					if (n) {
						let r = C.current[t], i = e[n];
						r && i instanceof re && i.addEventListener(r);
					}
				}
			}
			w.current = C.current;
		}
		s && y.current && (x.current = {
			...b,
			...s(y.current, b, v, O.current)
		}, ne());
		let r = _ ? x.current?.[J] : k;
		g && r && y.current && r.setEvents(_ ? null : y.current, C.current), j.current || te(!0);
	}, [b]), P = n(() => {
		te(!1), E.current = !1;
	}, []), F = n(async () => {
		await new Promise((e) => queueMicrotask(() => e(void 0))), A.current &&= (await A.current, void 0), y.current && r && r(y.current, b, D.current, O.current);
		let e = _ ? x.current?.[J] : k;
		if (g && e && y.current && e.clearEvents(_ ? null : y.current), y.current && !He(y.current)) {
			let e = Object.keys(S.current);
			for (let t of e) y.current[t]?.removeEventListener?.(S.current[t]);
		}
		S.current = {}, w.current = {}, x.current = void 0, O.current = void 0, y.current = void 0;
	}, [b]);
	return o(() => ((async () => {
		j.current &&= (await j.current, void 0), A.current = N();
	})(), () => {
		P(), j.current = F();
	}), [
		N,
		F,
		P
	]), i(() => {
		(async () => {
			A.current && await A.current;
			let e = Ge(v);
			T ? Be(e, w.current) || (await M(e), b.__$internal?.onUpdate?.()) : (w.current = e, C.current = e, E.current = !0);
		})();
	}, [
		b.__$internal,
		T,
		v,
		M
	]), a(ee, () => ({ cesiumElement: T ? y.current : null }), [T]), [
		x.current,
		T,
		D
	];
};
function Ge(e) {
	let { children: t, ...n } = e;
	return n;
}
//#endregion
//#region src/core/component.tsx
var Ke = () => void 0, X = ({ renderContainer: e, noChildren: n, containerProps: r, defaultProps: i, useResource: a = Ke, ...o }) => {
	let s = (t, s) => {
		let c = {
			...i,
			...t
		}, l = a(c), u = l ? {
			...c,
			...l
		} : c, [d, f, p] = We(o, u, s);
		if (n) return null;
		let m = f && "children" in u ? u.children : null, h = e ? /* @__PURE__ */ V("div", {
			"data-testid": "resium-container",
			ref: p,
			...typeof r == "function" ? r(u) : ze(u, r),
			children: m
		}) : m ? /* @__PURE__ */ V(B, { children: m }) : null;
		return d ? /* @__PURE__ */ V(H.Provider, {
			value: d,
			children: h
		}) : h;
	};
	return s.displayName = o.name, t(s);
}, qe = ["enabled", "selected"], Z = (e) => X({
	name: e.name,
	create(t, n) {
		if (!t.scene) return;
		let r = e.create(n, t.scene.postProcessStages);
		return typeof n.enabled == "boolean" && (r.enabled = n.enabled), n.selected && "selected" in r && (r.selected = n.selected), e.props.forEach((t) => {
			!K(e.readonlyProps, t) && n[t] !== void 0 && (r.uniforms[t] = n[t]);
		}), !e.noMount && t.scene && !t.scene.isDestroyed() && t.scene.postProcessStages.add(r), r;
	},
	destroy(t, n) {
		e.noMount ? t.enabled = !1 : (n.scene && !n.scene.isDestroyed() && n.scene.postProcessStages.remove(t), t.isDestroyed() || t.destroy());
	},
	update(t, n, r) {
		e.props.forEach((i) => {
			!K(e.readonlyProps, i) && n[i] !== r[i] && (t.uniforms[i] = n[i]);
		});
	},
	cesiumProps: qe,
	cesiumReadonlyProps: e.readonlyProps,
	defaultProps: { enabled: !0 }
}), Q = /* @__PURE__ */ new Map();
function Je(e, t) {
	if (!e) return;
	let n = Q.get(e);
	if (!n) {
		let r = {
			status: "pending",
			promise: Promise.resolve()
		};
		r.promise = t().then((e) => {
			r.status = "fulfilled", r.value = e;
		}, (e) => {
			r.status = "rejected", r.error = e;
		}), Q.set(e, r), n = r;
	}
	if (n.status === "pending") throw n.promise;
	if (n.status === "rejected") throw n.error;
	return n.value;
}
function Ye(e) {
	e ? Q.delete(e) : Q.clear();
}
function Xe(e) {
	if (typeof e == "string") return e;
	if (e instanceof L) return e.url;
}
function $(e, t, n, r) {
	let i = n.suspense ? Xe(t) : void 0;
	return Je(i ? `${e}:${n.cacheKey ?? i}` : void 0, () => r(i));
}
//#endregion
//#region src/core/withCesium.tsx
var Ze = (e) => t((t, n) => /* @__PURE__ */ V(Re, { children: (r) => /* @__PURE__ */ V(e, {
	...t,
	ref: n,
	cesium: r
}) })), Qe = X({
	name: "Billboard",
	create(e, t) {
		return e.billboardCollection?.add(t);
	},
	destroy(e, t) {
		t.billboardCollection && !t.billboardCollection.isDestroyed() && t.billboardCollection.remove(e);
	},
	cesiumProps: [
		"alignedAxis",
		"color",
		"disableDepthTestDistance",
		"distanceDisplayCondition",
		"eyeOffset",
		"height",
		"heightReference",
		"horizontalOrigin",
		"image",
		"pixelOffset",
		"pixelOffsetScaleByDistance",
		"position",
		"rotation",
		"scale",
		"scaleByDistance",
		"show",
		"sizeInMeters",
		"splitDirection",
		"translucencyByDistance",
		"verticalOrigin",
		"width",
		"id"
	],
	useCommonEvent: !0
}), $e = X({
	name: "BillboardCollection",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new d({
			modelMatrix: t.modelMatrix,
			debugShowBoundingVolume: t.debugShowBoundingVolume,
			scene: e.scene,
			blendOption: t.blendOption
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { billboardCollection: e };
	},
	cesiumProps: [
		"blendOption",
		"coarseDepthTestDistance",
		"debugShowBoundingVolume",
		"debugShowTextureAtlas",
		"modelMatrix",
		"show",
		"threePointDepthTestDistance"
	],
	setCesiumPropsAfterCreate: !0
}), et = X({
	name: "BillboardGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new f(t);
		return e.entity.billboard = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.billboard = void 0);
	},
	cesiumProps: [
		"image",
		"show",
		"scale",
		"horizontalOrigin",
		"verticalOrigin",
		"eyeOffset",
		"pixelOffset",
		"rotation",
		"alignedAxis",
		"width",
		"height",
		"color",
		"scaleByDistance",
		"translucencyByDistance",
		"pixelOffsetScaleByDistance",
		"imageSubRegion",
		"sizeInMeters",
		"heightReference",
		"distanceDisplayCondition",
		"disableDepthTestDistance",
		"splitDirection"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), tt = X({
	name: "BoxGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new p(t);
		return e.entity.box = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.box = void 0);
	},
	cesiumProps: [
		"heightReference",
		"dimensions",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"shadows",
		"distanceDisplayCondition"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), nt = X({
	name: "BufferPoint",
	create(e, t) {
		if (!e.bufferPointCollection) return;
		let n = new m(), r = e.bufferPointCollection.add({
			show: t.show,
			position: t.position,
			material: t.material
		}, n);
		return t.featureId !== void 0 && (r.featureId = t.featureId), r;
	},
	destroy(e) {
		e.show = !1;
	},
	update(e, t, n) {
		t.position !== n.position && t.position !== void 0 && e.setPosition(t.position), t.material !== n.material && t.material !== void 0 && e.setMaterial(t.material);
	},
	cesiumProps: ["show", "featureId"],
	otherProps: ["position", "material"]
}), rt = X({
	name: "BufferPointCollection",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new h({
			primitiveCountMax: t.primitiveCountMax,
			modelMatrix: t.modelMatrix,
			boundingVolume: t.boundingVolume,
			blendOption: t.blendOption
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { bufferPointCollection: e };
	},
	cesiumProps: ["show", "debugShowBoundingVolume"],
	cesiumReadonlyProps: [
		"primitiveCountMax",
		"modelMatrix",
		"boundingVolume",
		"blendOption"
	],
	setCesiumPropsAfterCreate: !0
}), it = X({
	name: "BufferPolygon",
	create(e, t) {
		if (!e.bufferPolygonCollection) return;
		let n = new g(), r = e.bufferPolygonCollection.add({
			show: t.show,
			positions: t.positions,
			holes: t.holes,
			triangles: t.triangles,
			material: t.material
		}, n);
		return t.featureId !== void 0 && (r.featureId = t.featureId), r;
	},
	destroy(e) {
		e.show = !1;
	},
	update(e, t, n) {
		t.positions !== n.positions && t.positions !== void 0 && e.setPositions(t.positions), t.holes !== n.holes && t.holes !== void 0 && e.setHoles(t.holes), t.triangles !== n.triangles && t.triangles !== void 0 && e.setTriangles(t.triangles), t.material !== n.material && t.material !== void 0 && e.setMaterial(t.material);
	},
	cesiumProps: ["show", "featureId"],
	otherProps: [
		"positions",
		"holes",
		"triangles",
		"material"
	]
}), at = X({
	name: "BufferPolygonCollection",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new _({
			primitiveCountMax: t.primitiveCountMax,
			vertexCountMax: t.vertexCountMax,
			holeCountMax: t.holeCountMax,
			triangleCountMax: t.triangleCountMax,
			positionDatatype: t.positionDatatype,
			allowPicking: t.allowPicking,
			modelMatrix: t.modelMatrix,
			boundingVolume: t.boundingVolume,
			blendOption: t.blendOption
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { bufferPolygonCollection: e };
	},
	cesiumProps: ["show", "debugShowBoundingVolume"],
	cesiumReadonlyProps: [
		"primitiveCountMax",
		"vertexCountMax",
		"holeCountMax",
		"triangleCountMax",
		"positionDatatype",
		"allowPicking",
		"modelMatrix",
		"boundingVolume",
		"blendOption"
	],
	setCesiumPropsAfterCreate: !0
}), ot = X({
	name: "BufferPolyline",
	create(e, t) {
		if (!e.bufferPolylineCollection) return;
		let n = new v(), r = e.bufferPolylineCollection.add({
			show: t.show,
			positions: t.positions,
			material: t.material
		}, n);
		return t.featureId !== void 0 && (r.featureId = t.featureId), r;
	},
	destroy(e) {
		e.show = !1;
	},
	update(e, t, n) {
		t.positions !== n.positions && t.positions !== void 0 && e.setPositions(t.positions), t.material !== n.material && t.material !== void 0 && e.setMaterial(t.material);
	},
	cesiumProps: ["show", "featureId"],
	otherProps: ["positions", "material"]
}), st = X({
	name: "BufferPolylineCollection",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new ee({
			primitiveCountMax: t.primitiveCountMax,
			vertexCountMax: t.vertexCountMax,
			modelMatrix: t.modelMatrix,
			boundingVolume: t.boundingVolume,
			blendOption: t.blendOption
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { bufferPolylineCollection: e };
	},
	cesiumProps: ["show", "debugShowBoundingVolume"],
	cesiumReadonlyProps: [
		"primitiveCountMax",
		"vertexCountMax",
		"modelMatrix",
		"boundingVolume",
		"blendOption"
	],
	setCesiumPropsAfterCreate: !0
}), ct = X({
	name: "Camera",
	create: (e) => e.scene?.camera,
	cesiumProps: [
		"position",
		"direction",
		"up",
		"right",
		"frustum",
		"defaultMoveAmount",
		"defaultLookAmount",
		"defaultRotateAmount",
		"defaultZoomAmount",
		"constrainedAxis",
		"maximumZoomFactor",
		"percentageChanged"
	],
	cesiumEventProps: {
		onChange: "changed",
		onMoveEnd: "moveEnd",
		onMoveStart: "moveStart"
	},
	setCesiumPropsAfterCreate: !0
}), lt = W("CameraFlyHome", (e, { duration: t }) => {
	e.flyHome(t);
}), ut = W("CameraFlyTo", (e, { onComplete: t, onCancel: n, ...r }) => {
	e.flyTo({
		...r,
		complete: t,
		cancel: n
	});
}), dt = W("CameraLookAt", (e, { target: t, offset: n }) => {
	e.lookAt(t, n);
}), ft = W("CameraFlyToBoundingSphere", (e, { boundingSphere: t, onComplete: n, onCancel: r, ...i }) => {
	e.flyToBoundingSphere(t, {
		...i,
		complete: n,
		cancel: r
	});
}), pt = /* @__PURE__ */ "show.modelMatrix.shadows.maximumScreenSpaceError.cullRequestsWhileMoving.cullRequestsWhileMovingMultiplier.preloadWhenHidden.preloadFlightDestinations.preferLeaves.progressiveResolutionHeightFraction.foveatedScreenSpaceError.foveatedConeSize.foveatedMinimumScreenSpaceErrorRelaxation.foveatedInterpolationCallback.foveatedTimeDelay.dynamicScreenSpaceError.dynamicScreenSpaceErrorDensity.dynamicScreenSpaceErrorFactor.dynamicScreenSpaceErrorHeightFalloff.edgeDisplayMode.skipLevelOfDetail.baseScreenSpaceError.skipScreenSpaceErrorFactor.skipLevels.immediatelyLoadDesiredLevelOfDetail.loadSiblings.clippingPlanes.clippingPolygons.lightColor.colorBlendAmount.colorBlendMode.debugFreezeFrame.debugColorizeTiles.debugWireframe.debugShowBoundingVolume.debugShowContentBoundingVolume.debugShowViewerRequestVolume.debugShowGeometricError.debugShowRenderingStatistics.debugShowMemoryUsage.debugShowUrl.style.backFaceCulling.showOutline.vectorClassificationOnly.vectorKeepDecodedPositions.splitDirection.customShader.imageBasedLighting.showCreditsOnScreen.featureIdLabel.instanceFeatureIdLabel.outlineColor.cacheBytes.maximumCacheOverflowBytes.enableCollision".split("."), mt = [
	"asynchronouslyLoadImagery",
	"classificationType",
	"cullWithChildrenBounds",
	"debugHeatmapTilePropertyName",
	"ellipsoid",
	"enableDebugWireframe",
	"heightReference",
	"modelUpAxis",
	"modelForwardAxis",
	"projectTo2D",
	"enableShowOutline",
	"enablePick",
	"environmentMapOptions",
	"scene"
], ht = {
	onAllTilesLoad: "allTilesLoaded",
	onInitialTilesLoad: "initialTilesLoaded",
	onLoadProgress: "loadProgress",
	onTileFailed: "tileFailed",
	onTileLoad: "tileLoad",
	onTileUnload: "tileUnload",
	onTileVisible: "tileVisible"
}, gt = ["onReady", "onError"], _t = [...mt, "url"], vt = (e) => {
	let t = e;
	t._clippingPlanes = void 0, t._clippingPolygons = void 0;
}, yt = X({
	name: "Cesium3DTileset",
	async create(e, t) {
		if (!e.primitiveCollection) return;
		let n = t.url, r;
		r = q(n) ? await n : n;
		let i;
		try {
			i = await b.fromUrl(r, t), t.onReady?.(i);
		} catch (e) {
			t.onError?.(e);
			return;
		}
		return t.colorBlendAmount && (i.colorBlendAmount = t.colorBlendAmount), t.colorBlendMode && (i.colorBlendMode = t.colorBlendMode), t.style && (i.style = t.style), e.primitiveCollection.add(i), i;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || (vt(e), e.destroy());
	},
	cesiumProps: pt,
	cesiumReadonlyProps: _t,
	cesiumEventProps: ht,
	otherProps: gt,
	useCommonEvent: !0
}), bt = X({
	name: "Cesium3DTilesetGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new x(t);
		return e.entity.tileset = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.tileset = void 0);
	},
	cesiumProps: [
		"show",
		"uri",
		"maximumScreenSpaceError"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), xt = X({
	name: "Cesium3DTilesTerrainProvider",
	async create(e, t) {
		let { onReady: n, assetId: r, accessToken: i, url: a, children: o, ...s } = t, c;
		if (r !== void 0) c = await y.fromIonAssetId(r, s);
		else if (a !== void 0) c = await y.fromUrl(a, s);
		else throw Error("Cesium3DTilesTerrainProvider requires either 'url' or 'assetId' prop to be provided");
		return n && n(c), c;
	}
}), St = X({
	name: "CesiumWidget",
	create(e, t, n) {
		if (!n) return;
		let r = new S(n, t);
		if (r) return typeof t.resolutionScale == "number" && (r.resolutionScale = t.resolutionScale), [r, new Ue(r.scene)];
	},
	destroy(e, t, n, r) {
		r && !r.isDestroyed() && r.destroy(), e.isDestroyed() || e.destroy();
	},
	provide(e, t, n, r) {
		return {
			cesiumWidget: e,
			scene: e.scene,
			camera: e.scene.camera,
			imageryLayerCollection: e.scene.globe.imageryLayers,
			primitiveCollection: e.scene.primitives,
			globe: e.scene.globe,
			__$internal: { onUpdate: n?.onUpdate },
			[J]: r
		};
	},
	containerProps: ({ id: e, className: t, style: n, full: r, containerProps: i }) => ({
		className: t,
		id: e,
		style: {
			...r ? {
				position: "absolute",
				bottom: "0",
				left: "0",
				right: "0",
				top: "0"
			} : {},
			...n
		},
		...i
	}),
	cesiumProps: [
		"resolutionScale",
		"useDefaultRenderLoop",
		"targetFrameRate",
		"useBrowserRecommendedResolution",
		"allowDataSourcesToSuspendAnimation",
		"trackedEntity",
		"clockTrackedDataSource",
		"terrainProvider",
		"creditDisplay"
	],
	cesiumReadonlyProps: [
		"clock",
		"shouldAnimate",
		"ellipsoid",
		"baseLayer",
		"skyBox",
		"skyAtmosphere",
		"sceneMode",
		"scene3DOnly",
		"orderIndependentTranslucency",
		"mapMode2D",
		"mapProjection",
		"globe",
		"showRenderLoopErrors",
		"automaticallyTrackDataSourceClocks",
		"contextOptions",
		"creditContainer",
		"creditViewport",
		"dataSources",
		"shadows",
		"terrainShadows",
		"terrain",
		"requestRenderMode",
		"maximumRenderTimeChange",
		"msaaSamples",
		"blurActiveElementOnCanvasFocus"
	],
	otherProps: [
		"className",
		"id",
		"style",
		"full",
		"containerProps"
	],
	renderContainer: !0,
	useCommonEvent: !0,
	useRootEvent: !0
}), Ct = X({
	name: "ClassificationPrimitive",
	async create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new C(t);
		if (t.onReady) {
			let r = () => {
				n.ready && (t.onReady?.(n), e.scene?.postRender.removeEventListener(r));
			};
			e.scene?.postRender.addEventListener(r);
		}
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: [
		"classificationType",
		"debugShowBoundingVolume",
		"debugShowShadowVolume",
		"show"
	],
	cesiumReadonlyProps: [
		"allowPicking",
		"asynchronous",
		"compressVertices",
		"geometryInstances",
		"interleave",
		"releaseGeometryInstances",
		"vertexCacheOptimize",
		"appearance"
	],
	otherProps: ["onReady"],
	useCommonEvent: !0
}), wt = X({
	name: "Clock",
	create: (e) => e.cesiumWidget?.clock,
	cesiumProps: [
		"canAnimate",
		"clockRange",
		"clockStep",
		"currentTime",
		"multiplier",
		"shouldAnimate",
		"startTime",
		"stopTime"
	],
	cesiumEventProps: {
		onStop: "onStop",
		onTick: "onTick"
	},
	setCesiumPropsAfterCreate: !0
}), Tt = X({
	name: "CloudCollection",
	create: (e) => {
		if (!e.primitiveCollection) return;
		let t = new w();
		return e.primitiveCollection.add(t), t;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide: (e) => ({ cloudCollection: e }),
	cesiumProps: [
		"noiseDetail",
		"noiseOffset",
		"show",
		"debugBillboards",
		"debugEllipsoids"
	],
	setCesiumPropsAfterCreate: !0
}), Et = X({
	name: "CorridorGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new te(t);
		return t.classificationType && (n.classificationType = t.classificationType), e.entity.corridor = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.corridor = void 0);
	},
	cesiumProps: [
		"positions",
		"width",
		"cornerType",
		"height",
		"heightReference",
		"extrudedHeight",
		"extrudedHeightReference",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"granularity",
		"shadows",
		"distanceDisplayCondition",
		"zIndex",
		"classificationType"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), Dt = X({
	name: "CubeMapPanorama",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new E({
			sources: t.sources,
			show: t.show,
			transform: t.transform,
			credit: t.credit
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	setCesiumPropsAfterCreate: !0,
	cesiumProps: ["show", "sources"],
	cesiumReadonlyProps: ["transform", "credit"]
}), Ot = X({
	name: "CumulusCloud",
	create: (e, t) => e.cloudCollection?.add(t),
	destroy(e, t) {
		t.cloudCollection && !t.cloudCollection.isDestroyed() && t.cloudCollection.remove(e);
	},
	cesiumProps: [
		"show",
		"position",
		"scale",
		"maximumSize",
		"slice",
		"brightness",
		"color"
	]
}), kt = X({
	name: "CustomDataSource",
	create(e, t) {
		if (!e.dataSourceCollection) return;
		let n = new ne(t.name);
		return t.clustering && (n.clustering = t.clustering), typeof t.show == "boolean" && (n.show = t.show), t.clock !== void 0 && (n.clock = t.clock), e.dataSourceCollection.add(n), n;
	},
	destroy(e, t) {
		t.dataSourceCollection && !t.dataSourceCollection.isDestroyed() && t.dataSourceCollection.remove(e);
	},
	provide(e) {
		return {
			entityCollection: e.entities,
			dataSource: e
		};
	},
	cesiumProps: [
		"clustering",
		"name",
		"show",
		"clock",
		"isLoading"
	],
	cesiumEventProps: {
		onChange: "changedEvent",
		onError: "errorEvent",
		onLoading: "loadingEvent"
	},
	useCommonEvent: !0
}), At = X({
	name: "CylinderGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new D(t);
		return e.entity.cylinder = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.cylinder = void 0);
	},
	cesiumProps: [
		"heightReference",
		"length",
		"topRadius",
		"bottomRadius",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"numberOfVerticalLines",
		"slices",
		"distanceDisplayCondition",
		"shadows"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), jt = ["clustering", "show"], Mt = [
	"name",
	"sourceUri",
	"credit"
], Nt = {
	onChange: "changedEvent",
	onError: "errorEvent",
	onLoading: "loadingEvent"
}, Pt = [
	"onLoad",
	"data",
	"suspense",
	"cacheKey"
], Ft = (e) => {
	let t = $("czml", e.data, e, (e) => L.fetchJson({ url: e }));
	return t ? { data: t } : void 0;
}, It = (e, { data: t, onLoad: n, ...r }) => {
	t && e.load(t, r).then((e) => {
		n && n(e);
	});
}, Lt = X({
	name: "CzmlDataSource",
	create(e, t) {
		if (!e.dataSourceCollection) return;
		let n = new O(t.name);
		return t.clustering && (n.clustering = t.clustering), typeof t.show == "boolean" && (n.show = t.show), e.dataSourceCollection.add(n), t.data && It(n, t), n;
	},
	update(e, t, n) {
		t.data ? n.show !== t.show && (e.show = typeof t.show == "boolean" ? t.show : !0) : e.show = !1, t.data && (n.data !== t.data || n.sourceUri !== t.sourceUri || n.credit !== t.credit) && It(e, t);
	},
	destroy(e, t) {
		t.dataSourceCollection && !t.dataSourceCollection.isDestroyed() && t.dataSourceCollection.remove(e);
	},
	provide(e) {
		return { dataSource: e };
	},
	useResource: Ft,
	cesiumProps: jt,
	cesiumReadonlyProps: Mt,
	cesiumEventProps: Nt,
	otherProps: Pt,
	useCommonEvent: !0
}), Rt = X({
	name: "EllipseGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new k(t);
		return e.entity.ellipse = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.ellipse = void 0);
	},
	cesiumProps: [
		"semiMajorAxis",
		"semiMinorAxis",
		"height",
		"heightReference",
		"extrudedHeight",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"numberOfVerticalLines",
		"rotation",
		"stRotation",
		"granularity",
		"shadows",
		"distanceDisplayCondition",
		"zIndex",
		"classificationType",
		"extrudedHeightReference"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), zt = X({
	name: "EllipsoidGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new A(t);
		return e.entity.ellipsoid = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.ellipsoid = void 0);
	},
	cesiumProps: [
		"heightReference",
		"radii",
		"show",
		"fill",
		"innerRadii",
		"material",
		"maximumClock",
		"maximumCone",
		"minimumClock",
		"minimumCone",
		"outline",
		"outlineColor",
		"outlineWidth",
		"subdivisions",
		"stackPartitions",
		"slicePartitions",
		"shadows",
		"distanceDisplayCondition"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), Bt = X({
	name: "Entity",
	create(e, t) {
		if (!e.entityCollection) return;
		let n = new j(t);
		return e.viewer && t.selected && (e.viewer.selectedEntity = n), e.viewer && t.tracked && (e.viewer.trackedEntity = n), e.entityCollection.add(n), n;
	},
	destroy(e, t) {
		t.entityCollection && t.entityCollection.remove(e);
	},
	update(e, t, n, r) {
		r.viewer && (t.selected !== n.selected && (t.selected ? r.viewer.selectedEntity = e : r.viewer.selectedEntity === e && (r.viewer.selectedEntity = void 0)), t.tracked !== n.tracked && (t.tracked ? r.viewer.trackedEntity = e : r.viewer.trackedEntity === e && (r.viewer.trackedEntity = void 0)));
	},
	provide(e, t, n) {
		return {
			entity: e,
			__$internal: { onUpdate: n?.onUpdate }
		};
	},
	cesiumProps: /* @__PURE__ */ "availability.billboard.box.corridor.cylinder.description.ellipse.ellipsoid.entityCollection.label.model.name.orientation.path.plane.parent.point.polygon.polyline.polylineVolume.position.properties.rectangle.show.tileset.trackingReferenceFrame.viewFrom.wall".split("."),
	cesiumReadonlyProps: ["id"],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" },
	otherProps: ["selected", "tracked"],
	useCommonEvent: !0
}), Vt = X({
	name: "EquirectangularPanorama",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new M({
			transform: t.transform,
			image: t.image,
			radius: t.radius,
			repeatHorizontal: t.repeatHorizontal,
			repeatVertical: t.repeatVertical,
			credit: t.credit
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	setCesiumPropsAfterCreate: !0,
	cesiumProps: ["show"],
	cesiumReadonlyProps: [
		"transform",
		"image",
		"radius",
		"repeatHorizontal",
		"repeatVertical",
		"credit"
	]
}), Ht = ({ children: e, container: t, resizeInfoBox: n = !0 }) => {
	let { viewer: r, entity: a } = U(), [o, c] = u(!1), l = s(() => t ?? r?.infoBox.frame.contentDocument?.createElement("div"), [t, r?.infoBox.frame.contentDocument]);
	return i(() => {
		if (!r || !a) return;
		let e = (e) => {
			c(!!e && e.id === a.id);
		};
		return r.selectedEntityChanged.addEventListener(e), () => {
			r.selectedEntityChanged.removeEventListener(e);
		};
	}, [a, r]), i(() => {
		if (t || !l || !r) return;
		let e = r.infoBox?.frame, i = e?.contentDocument?.querySelector(".cesium-infoBox-description");
		if (!e || !i) return;
		let a;
		if (o) {
			if (n) {
				let t = i.getBoundingClientRect().height;
				e.style.height = t + "px", a = window.setInterval(() => {
					let t = r.infoBox.container.querySelector(".cesium-infoBox.cesium-infoBox-visible");
					t && (clearInterval(a), a = void 0, i.appendChild(l), t.classList.remove("cesium-infoBox-bodyless"), e.style.height = i.getBoundingClientRect().height + "px");
				}, 10);
			}
		} else l.parentElement === i && i.removeChild(l);
		return a ? () => clearTimeout(a) : void 0;
	}, [
		l,
		t,
		n,
		o,
		r
	]), l ? /* @__PURE__ */ V(B, { children: Ie(/* @__PURE__ */ V(B, { children: !t || o ? e : null }), l) }) : null;
}, Ut = X({
	name: "Fog",
	create(e) {
		if (!e.scene) return;
		let t = new N();
		return e.scene.fog = t, t;
	},
	destroy(e, t) {
		t.scene && !t.scene.isDestroyed() && (t.scene.fog = new N());
	},
	cesiumProps: [
		"density",
		"enabled",
		"heightFalloff",
		"heightScalar",
		"maxHeight",
		"minimumBrightness",
		"renderable",
		"screenSpaceErrorFactor",
		"visualDensityScalar"
	],
	setCesiumPropsAfterCreate: !0
}), Wt = [
	"clustering",
	"name",
	"show"
], Gt = [
	"clampToGround",
	"sourceUri",
	"credit",
	"markerSize",
	"markerSymbol",
	"markerColor",
	"stroke",
	"strokeWidth",
	"fill",
	"describe"
], Kt = {
	onChange: "changedEvent",
	onError: "errorEvent",
	onLoading: "loadingEvent"
}, qt = [
	"onLoad",
	"data",
	"suspense",
	"cacheKey"
], Jt = (e) => {
	let t = $("geojson", e.data, e, (e) => L.fetchJson({ url: e }));
	return t ? { data: t } : void 0;
}, Yt = (e, { data: t, onLoad: n, ...r }) => {
	t && e.load(t, r).then((e) => {
		n && n(e);
	});
}, Xt = X({
	name: "GeoJsonDataSource",
	create(e, t) {
		if (!e.dataSourceCollection) return;
		let n = new P(t.name);
		return t.clustering && (n.clustering = t.clustering), typeof t.show == "boolean" && (n.show = t.show), e.dataSourceCollection.add(n), t.data && Yt(n, t), n;
	},
	update(e, t, n) {
		t.data ? n.show !== t.show && (e.show = typeof t.show == "boolean" ? t.show : !0) : e.show = !1, t.data && (n.data !== t.data || n.clampToGround !== t.clampToGround || n.sourceUri !== t.sourceUri || n.credit !== t.credit || n.markerSize !== t.markerSize || n.markerSymbol !== t.markerSymbol || n.markerColor !== t.markerColor || n.stroke !== t.stroke || n.strokeWidth !== t.strokeWidth || n.fill !== t.fill) && Yt(e, t);
	},
	destroy(e, t) {
		t.dataSourceCollection && !t.dataSourceCollection.isDestroyed() && t.dataSourceCollection.remove(e);
	},
	provide(e) {
		return { dataSource: e };
	},
	useResource: Jt,
	cesiumProps: Wt,
	cesiumReadonlyProps: Gt,
	cesiumEventProps: Kt,
	otherProps: qt,
	useCommonEvent: !0
}), Zt = X({
	name: "GeoJsonPrimitive",
	async create(e, t) {
		if (!e.primitiveCollection) return;
		let { url: n, data: r, ellipsoid: i, allowPicking: a, pickObjectFactory: o, show: s, onReady: c, onError: l } = t, u;
		try {
			if (n) u = await F.fromUrl(n, {
				ellipsoid: i,
				allowPicking: a,
				pickObjectFactory: o,
				show: s
			});
			else if (r) u = new F({
				geoJson: r,
				ellipsoid: i,
				allowPicking: a,
				pickObjectFactory: o,
				show: s
			});
			else return;
			c?.(u);
		} catch (e) {
			l?.(e);
			return;
		}
		return e.primitiveCollection.add(u), u;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e);
	},
	useResource: (e) => {
		let t = $("geojson-primitive", e.url ?? e.data, e, (e) => L.fetchJson({ url: e }));
		return t ? {
			data: t,
			url: void 0
		} : void 0;
	},
	cesiumProps: ["show"],
	cesiumReadonlyProps: [
		"ellipsoid",
		"allowPicking",
		"pickObjectFactory",
		"url",
		"data"
	],
	otherProps: [
		"url",
		"data",
		"onReady",
		"onError",
		"suspense",
		"cacheKey"
	],
	useCommonEvent: !0
}), Qt = X({
	name: "Globe",
	create: (e) => e.scene?.globe,
	update: async (e, t, n) => {
		if (t.terrainProvider === n.terrainProvider) return;
		let r = t.terrainProvider, i;
		i = q(r) ? await r : r, e.terrainProvider = i;
	},
	cesiumProps: /* @__PURE__ */ "atmosphereBrightnessShift.atmosphereHueShift.atmosphereSaturationShift.backFaceCulling.baseColor.clippingPlanes.depthTestAgainstTerrain.enableLighting.lightingFadeInDistance.lightingFadeOutDistance.material.maximumScreenSpaceError.nightFadeInDistance.nightFadeOutDistance.oceanNormalMapUrl.shadows.show.showGroundAtmosphere.showWaterEffect.tileCacheSize.loadingDescendantLimit.preloadAncestors.preloadSiblings.fillHighlightColor.dynamicAtmosphereLighting.dynamicAtmosphereLightingFromSun.showSkirts.cartographicLimitRectangle.translucency.undergroundColor.undergroundColorAlphaByDistance.lambertDiffuseMultiplier.atmosphereLightIntensity.atmosphereRayleighCoefficient.atmosphereMieCoefficient.atmosphereRayleighScaleHeight.atmosphereMieScaleHeight.atmosphereMieAnisotropy.vertexShadowDarkness.clippingPolygons".split("."),
	cesiumEventProps: {
		onImageryLayersUpdate: "imageryLayersUpdatedEvent",
		onTerrainProviderChange: "terrainProviderChanged",
		onTileLoadProgress: "tileLoadProgressEvent"
	},
	otherProps: ["terrainProvider"],
	setCesiumPropsAfterCreate: !0
}), $t = X({
	name: "Google2DImageryProvider",
	async create(e, t) {
		let { onReady: n, assetId: r, accessToken: i, key: a, children: o, ...s } = t, c;
		if (r !== void 0) {
			let e = {
				...s,
				assetId: r
			};
			i !== void 0 && (e.accessToken = i), c = await ie.fromIonAssetId(e);
		} else {
			let e = { ...s };
			a !== void 0 && (e.key = a), c = await ie.fromUrl(e);
		}
		return n && n(c), c;
	}
}), en = X({
	name: "GooglePhotorealistic3DTileset",
	async create(e, t) {
		if (!e.primitiveCollection) return;
		let { apiKey: n, onlyUsingWithGoogleGeocoder: r } = t, i = {};
		n !== void 0 && (i.key = n), r !== void 0 && (i.onlyUsingWithGoogleGeocoder = r);
		let a;
		try {
			a = await Fe(i, t), t.onReady?.(a);
		} catch (e) {
			t.onError?.(e);
			return;
		}
		return t.colorBlendAmount && (a.colorBlendAmount = t.colorBlendAmount), t.colorBlendMode && (a.colorBlendMode = t.colorBlendMode), t.style && (a.style = t.style), e.primitiveCollection.add(a), a;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: /* @__PURE__ */ "show.modelMatrix.shadows.maximumScreenSpaceError.cullRequestsWhileMoving.cullRequestsWhileMovingMultiplier.preloadWhenHidden.preloadFlightDestinations.preferLeaves.progressiveResolutionHeightFraction.foveatedScreenSpaceError.foveatedConeSize.foveatedMinimumScreenSpaceErrorRelaxation.foveatedInterpolationCallback.foveatedTimeDelay.dynamicScreenSpaceError.dynamicScreenSpaceErrorDensity.dynamicScreenSpaceErrorFactor.dynamicScreenSpaceErrorHeightFalloff.edgeDisplayMode.skipLevelOfDetail.baseScreenSpaceError.skipScreenSpaceErrorFactor.skipLevels.immediatelyLoadDesiredLevelOfDetail.loadSiblings.clippingPlanes.clippingPolygons.lightColor.colorBlendAmount.colorBlendMode.debugFreezeFrame.debugColorizeTiles.debugWireframe.debugShowBoundingVolume.debugShowContentBoundingVolume.debugShowViewerRequestVolume.debugShowGeometricError.debugShowRenderingStatistics.debugShowMemoryUsage.debugShowUrl.style.backFaceCulling.showOutline.vectorClassificationOnly.vectorKeepDecodedPositions.splitDirection.customShader.imageBasedLighting.showCreditsOnScreen.featureIdLabel.instanceFeatureIdLabel.outlineColor.cacheBytes.maximumCacheOverflowBytes.enableCollision".split("."),
	cesiumReadonlyProps: [
		"asynchronouslyLoadImagery",
		"classificationType",
		"cullWithChildrenBounds",
		"debugHeatmapTilePropertyName",
		"ellipsoid",
		"enableDebugWireframe",
		"heightReference",
		"modelUpAxis",
		"modelForwardAxis",
		"projectTo2D",
		"enableShowOutline",
		"enablePick",
		"environmentMapOptions",
		"scene"
	],
	cesiumEventProps: {
		onAllTilesLoad: "allTilesLoaded",
		onInitialTilesLoad: "initialTilesLoaded",
		onLoadProgress: "loadProgress",
		onTileFailed: "tileFailed",
		onTileLoad: "tileLoad",
		onTileUnload: "tileUnload",
		onTileVisible: "tileVisible"
	},
	otherProps: [
		"onReady",
		"onError",
		"apiKey",
		"onlyUsingWithGoogleGeocoder"
	],
	useCommonEvent: !0
}), tn = X({
	name: "GroundPolylinePrimitive",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new ae(t);
		if (t.onReady) {
			let r = () => {
				n.ready && (t.onReady?.(n), e.scene?.postRender.removeEventListener(r));
			};
			e.scene?.postRender.addEventListener(r);
		}
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: [
		"appearance",
		"classificationType",
		"debugShowBoundingVolume",
		"debugShowShadowVolume",
		"show"
	],
	cesiumReadonlyProps: [
		"allowPicking",
		"asynchronous",
		"geometryInstances",
		"interleave",
		"releaseGeometryInstances"
	],
	otherProps: ["onReady"],
	useCommonEvent: !0
}), nn = X({
	name: "GroundPrimitive",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new oe(t);
		if (t.onReady) {
			let r = () => {
				n.ready && (t.onReady?.(n), e.scene?.postRender.removeEventListener(r));
			};
			e.scene?.postRender.addEventListener(r);
		}
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: [
		"appearance",
		"classificationType",
		"debugShowBoundingVolume",
		"debugShowShadowVolume",
		"show"
	],
	cesiumReadonlyProps: [
		"allowPicking",
		"asynchronous",
		"compressVertices",
		"geometryInstances",
		"interleave",
		"releaseGeometryInstances",
		"vertexCacheOptimize"
	],
	otherProps: ["onReady"],
	useCommonEvent: !0
}), rn = X({
	name: "GroundPrimitiveCollection",
	create: (e) => e.scene?.groundPrimitives,
	provide: (e) => ({ primitiveCollection: e }),
	cesiumProps: [
		"show",
		"destroyPrimitives",
		"primitiveAdded",
		"primitiveRemoved"
	],
	setCesiumPropsAfterCreate: !0
}), an = X({
	name: "ImageryLayer",
	async create(e, t) {
		if (!e.imageryLayerCollection) return;
		let n = q(t.imageryProvider) ? t.imageryProvider : new Promise((e) => queueMicrotask(() => e(t.imageryProvider))), r = e.__$internal?.imageryLayerWaitingList?.slice();
		e.__$internal?.imageryLayerWaitingList && e.__$internal.imageryLayerWaitingList.push(n), r && await Promise.all(r.filter((e) => q(e)));
		let i = await n;
		if (e.__$internal?.imageryLayerWaitingList && (e.__$internal.imageryLayerWaitingList = e.__$internal.imageryLayerWaitingList.filter((e) => e !== n)), !i) return;
		let a = new se(i, t);
		return e.imageryLayerCollection.add(a, t.index), a;
	},
	destroy(e, t) {
		t.imageryLayerCollection && t.imageryLayerCollection.remove(e);
	},
	cesiumProps: [
		"alpha",
		"brightness",
		"contrast",
		"hue",
		"saturation",
		"gamma",
		"splitDirection",
		"minificationFilter",
		"magnificationFilter",
		"cutoutRectangle",
		"show",
		"nightAlpha",
		"dayAlpha",
		"colorToAlpha",
		"colorToAlphaThreshold",
		"index"
	],
	cesiumReadonlyProps: [
		"rectangle",
		"maximumAnisotropy",
		"minimumTerrainLevel",
		"maximumTerrainLevel",
		"imageryProvider"
	]
}), on = X({
	name: "ImageryLayerCollection",
	create: (e) => e.globe?.imageryLayers,
	cesiumEventProps: {
		onLayerAdd: "layerAdded",
		onLayerMove: "layerMoved",
		onLayerRemove: "layerRemoved",
		onLayerShowOrHide: "layerShownOrHidden"
	}
}), sn = [
	"clustering",
	"name",
	"show"
], cn = [
	"canvas",
	"camera",
	"ellipsoid",
	"clampToGround",
	"sourceUri",
	"credit",
	"screenOverlayContainer"
], ln = {
	onChange: "changedEvent",
	onError: "errorEvent",
	onLoading: "loadingEvent",
	onRefresh: "refreshEvent",
	onUnsupportedNode: "unsupportedNodeEvent"
}, un = [
	"onLoad",
	"data",
	"suspense",
	"cacheKey"
], dn = (e) => {
	let t = $("kml", e.data, e, (e) => L.fetchBlob({ url: e }));
	return t ? { data: t } : void 0;
}, fn = (e, { data: t, onLoad: n, ...r }) => {
	t && e.load(t, r).then((e) => {
		n && n(e);
	});
}, pn = X({
	name: "KmlDataSource",
	create(e, t) {
		if (!e.scene || !e.dataSourceCollection || !e.scene) return;
		let n = new ce({
			camera: t.camera || e.scene.camera,
			canvas: t.canvas || e.scene.canvas,
			ellipsoid: t.ellipsoid,
			credit: t.credit
		});
		return t.clustering && (n.clustering = t.clustering), typeof t.show == "boolean" && (n.show = t.show), t.name !== void 0 && (n.name = t.name), e.dataSourceCollection.add(n), t.data && fn(n, t), n;
	},
	update(e, t, n) {
		t.data ? n.show !== t.show && (e.show = typeof t.show == "boolean" ? t.show : !0) : e.show = !1, t.data && (n.data !== t.data || n.clampToGround !== t.clampToGround || n.ellipsoid !== t.ellipsoid || n.sourceUri !== t.sourceUri || n.credit !== t.credit) && fn(e, t);
	},
	destroy(e, t) {
		t.dataSourceCollection && !t.dataSourceCollection.isDestroyed() && t.dataSourceCollection.remove(e);
	},
	provide(e) {
		return { dataSource: e };
	},
	useResource: dn,
	cesiumProps: sn,
	cesiumReadonlyProps: cn,
	cesiumEventProps: ln,
	otherProps: un,
	useCommonEvent: !0
}), mn = X({
	name: "Label",
	create: (e, t) => e.labelCollection?.add(t),
	destroy(e, t) {
		t.labelCollection && !t.labelCollection.isDestroyed() && t.labelCollection.remove(e);
	},
	cesiumProps: [
		"backgroundColor",
		"backgroundPadding",
		"disableDepthTestDistance",
		"distanceDisplayCondition",
		"eyeOffset",
		"fillColor",
		"font",
		"heightReference",
		"horizontalOrigin",
		"id",
		"outlineColor",
		"outlineWidth",
		"pixelOffset",
		"pixelOffsetScaleByDistance",
		"position",
		"scale",
		"scaleByDistance",
		"show",
		"showBackground",
		"style",
		"text",
		"translucencyByDistance",
		"verticalOrigin"
	],
	useCommonEvent: !0
}), hn = X({
	name: "LabelCollection",
	create(e, t) {
		if (!e.scene || !e.primitiveCollection) return;
		let n = new le({
			scene: e.scene,
			modelMatrix: t.modelMatrix,
			blendOption: t.blendOption,
			debugShowBoundingVolume: t.debugShowBoundingVolume
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { labelCollection: e };
	},
	cesiumProps: [
		"blendOption",
		"coarseDepthTestDistance",
		"debugShowBoundingVolume",
		"modelMatrix",
		"show",
		"threePointDepthTestDistance"
	],
	setCesiumPropsAfterCreate: !0
}), gn = X({
	name: "LabelGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new ue(t);
		return e.entity.label = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.label = void 0);
	},
	cesiumProps: [
		"text",
		"font",
		"style",
		"fillColor",
		"outlineColor",
		"outlineWidth",
		"show",
		"showBackground",
		"backgroundColor",
		"backgroundPadding",
		"scale",
		"horizontalOrigin",
		"verticalOrigin",
		"eyeOffset",
		"pixelOffset",
		"translucencyByDistance",
		"pixelOffsetScaleByDistance",
		"scaleByDistance",
		"heightReference",
		"distanceDisplayCondition",
		"disableDepthTestDistance"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), _n = X({
	name: "Model",
	async create(e, { scene: t, url: n, colorBlendMode: r, ...i }) {
		if (!e.scene || !e.primitiveCollection || !n) return;
		let a = n, o;
		o = q(a) ? await a : a;
		let s;
		try {
			s = await pe.fromGltfAsync({
				...i,
				url: o,
				colorBlendMode: r,
				scene: t || e.scene
			});
		} catch (e) {
			i.onError?.(e);
			return;
		}
		return e.primitiveCollection.add(s), s;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || (e._clippingPlanes = void 0, e._clippingPolygons = void 0, e.destroy());
	},
	cesiumEventProps: { onReady: "readyEvent" },
	cesiumProps: /* @__PURE__ */ "backFaceCulling.clampAnimations.clippingPlanes.clippingPolygons.color.colorBlendAmount.colorBlendMode.customShader.debugShowBoundingVolume.debugWireframe.distanceDisplayCondition.edgeDisplayMode.enableVerticalExaggeration.featureIdLabel.heightReference.id.imageBasedLighting.instanceFeatureIdLabel.lightColor.maximumScale.minimumPixelSize.modelMatrix.outlineColor.scale.shadows.show.showCreditsOnScreen.showOutline.silhouetteColor.silhouetteSize.splitDirection.style.pointCloudShading".split("."),
	cesiumReadonlyProps: [
		"allowPicking",
		"asynchronous",
		"basePath",
		"credit",
		"enableDebugWireframe",
		"gltf",
		"incrementallyLoadTextures",
		"scene",
		"releaseGltfJson",
		"cull",
		"opaquePass",
		"upAxis",
		"forwardAxis",
		"content",
		"enableShowOutline",
		"projectTo2D",
		"classificationType",
		"gltfCallback",
		"enablePick",
		"environmentMapOptions"
	],
	otherProps: [
		"onReady",
		"onError",
		"url"
	],
	useCommonEvent: !0
}), vn = X({
	name: "ModelGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new me(t);
		return e.entity.model = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.model = void 0);
	},
	cesiumProps: [
		"uri",
		"show",
		"scale",
		"minimumPixelSize",
		"maximumScale",
		"incrementallyLoadTextures",
		"runAnimations",
		"clampAnimations",
		"nodeTransformations",
		"shadows",
		"heightReference",
		"distanceDisplayCondition",
		"silhouetteColor",
		"silhouetteSize",
		"color",
		"colorBlendMode",
		"colorBlendAmount",
		"clippingPlanes",
		"imageBasedLightingFactor",
		"lightColor",
		"articulations",
		"customShader",
		"enableVerticalExaggeration",
		"environmentMapOptions"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), yn = X({
	name: "Moon",
	create(e, t) {
		if (!e.scene) return;
		let n = new he(t);
		return e.scene.moon = n, n;
	},
	destroy(e, t) {
		t.scene && !t.scene.isDestroyed() && (t.scene.moon = new he());
	},
	cesiumProps: [
		"onlySunLighting",
		"show",
		"textureUrl"
	],
	cesiumReadonlyProps: ["ellipsoid"]
}), bn = X({
	name: "MVTDataProvider",
	async create(e, t) {
		if (!e.primitiveCollection) return;
		let { url: n, minZoom: r, maxZoom: i, extent: a, featureIdProperty: o, onReady: s, onError: c } = t, l;
		try {
			l = await de.fromUrl(n, {
				minZoom: r,
				maxZoom: i,
				extent: a,
				featureIdProperty: o
			}), s?.(l);
		} catch (e) {
			c?.(e);
			return;
		}
		return e.primitiveCollection.add(l), l;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: ["show"],
	cesiumReadonlyProps: [
		"minZoom",
		"maxZoom",
		"extent",
		"featureIdProperty",
		"url"
	],
	otherProps: [
		"url",
		"onReady",
		"onError"
	]
}), xn = (e) => e.tileset, Sn = X({
	name: "ParticleSystem",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new ge({
			...t,
			updateCallback: t.onUpdate
		});
		return e.primitiveCollection.add(n), n;
	},
	update(e, t, n) {
		t.onUpdate !== n.onUpdate && (e.updateCallback = t.onUpdate);
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e);
	},
	cesiumProps: /* @__PURE__ */ "show.emitter.modelMatrix.emitterModelMatrix.emissionRate.bursts.loop.startScale.endScale.startColor.endColor.image.imageSize.minimumImageSize.maximumImageSize.speed.minimumSpeed.maximumSpeed.lifetime.particleLife.minimumParticleLife.maximumParticleLife.mass.minimumMass.maximumMass.sizeInMeters".split("."),
	cesiumEventProps: {
		onComplete: "complete",
		onUpdate: "updateCallback"
	}
}), Cn = X({
	name: "PathGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new _e(t);
		return e.entity.path = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.path = void 0);
	},
	cesiumProps: [
		"leadTime",
		"trailTime",
		"show",
		"width",
		"material",
		"resolution",
		"distanceDisplayCondition",
		"relativeTo"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), wn = X({
	name: "PlaneGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new ve(t);
		return e.entity.plane = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.plane = void 0);
	},
	cesiumProps: [
		"plane",
		"dimensions",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"shadows",
		"distanceDisplayCondition"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), Tn = X({
	name: "PointGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new ye(t);
		return e.entity.point = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.point = void 0);
	},
	cesiumProps: [
		"color",
		"pixelSize",
		"outlineColor",
		"outlineWidth",
		"show",
		"scaleByDistance",
		"translucencyByDistance",
		"heightReference",
		"distanceDisplayCondition",
		"disableDepthTestDistance",
		"splitDirection"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), En = X({
	name: "PointPrimitive",
	create: (e, t) => e.pointPrimitiveCollection?.add(t),
	destroy(e, t) {
		t.pointPrimitiveCollection && !t.pointPrimitiveCollection.isDestroyed() && t.pointPrimitiveCollection.remove(e);
	},
	cesiumProps: [
		"color",
		"disableDepthTestDistance",
		"distanceDisplayCondition",
		"id",
		"outlineColor",
		"outlineWidth",
		"pixelSize",
		"position",
		"scaleByDistance",
		"show",
		"splitDirection",
		"translucencyByDistance"
	],
	useCommonEvent: !0
}), Dn = X({
	name: "PointPrimitveCollection",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new be(t);
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { pointPrimitiveCollection: e };
	},
	cesiumProps: [
		"blendOption",
		"debugShowBoundingVolume",
		"modelMatrix",
		"show"
	]
}), On = X({
	name: "PolygonGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new xe(t);
		return e.entity.polygon = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.polygon = void 0);
	},
	cesiumProps: [
		"arcType",
		"hierarchy",
		"height",
		"heightReference",
		"extrudedHeight",
		"extrudedHeightReference",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"stRotation",
		"granularity",
		"perPositionHeight",
		"closeTop",
		"closeBottom",
		"shadows",
		"distanceDisplayCondition",
		"zIndex",
		"classificationType",
		"textureCoordinates"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), kn = X({
	name: "Polyline",
	create: (e, t) => {
		let n = e.polylineCollection?.add(t);
		if (n) return [n, { userMaterial: !!t.material }];
	},
	destroy(e, t, n, r) {
		if (t.polylineCollection && !t.polylineCollection.isDestroyed()) {
			if (r?.userMaterial) {
				let t = e.material;
				t && !t.isDestroyed() && (e.material = fe.fromType(fe.ColorType, { color: new T(1, 1, 1, 1) }));
			}
			t.polylineCollection.remove(e);
		}
	},
	cesiumProps: [
		"distanceDisplayCondition",
		"id",
		"loop",
		"material",
		"positions",
		"show",
		"width"
	],
	useCommonEvent: !0
}), An = X({
	name: "PolylineCollection",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new Se({
			modelMatrix: t.modelMatrix,
			debugShowBoundingVolume: t.debugShowBoundingVolume,
			length: t.length,
			scene: e.scene
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { polylineCollection: e };
	},
	cesiumProps: [
		"debugShowBoundingVolume",
		"length",
		"modelMatrix",
		"show"
	]
}), jn = X({
	name: "PolylineGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new Ce(t);
		return e.entity.polyline = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.polyline = void 0);
	},
	cesiumProps: [
		"arcType",
		"classificationType",
		"positions",
		"clampToGround",
		"width",
		"show",
		"material",
		"depthFailMaterial",
		"granularity",
		"shadows",
		"distanceDisplayCondition",
		"zIndex"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), Mn = X({
	name: "PolylineVolumeGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new we(t);
		return e.entity.polylineVolume = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.polylineVolume = void 0);
	},
	cesiumProps: [
		"positions",
		"shape",
		"cornerType",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"granularity",
		"shadows",
		"distanceDisplayCondition"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), Nn = X({
	name: "PostProcessStage",
	create(e, t) {
		if (!e.scene) return;
		let n = new Te(t);
		return typeof t.enabled == "boolean" && (n.enabled = t.enabled), t.selected && (n.selected = t.selected), e.scene.postProcessStages.add(n), n;
	},
	destroy(e, t) {
		t.scene && !t.scene.isDestroyed() && t.scene.postProcessStages.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: ["enabled", "selected"],
	cesiumReadonlyProps: [
		"clearColor",
		"forcePowerOfTwo",
		"fragmentShader",
		"name",
		"pixelDatatype",
		"pixelFormat",
		"sampleMode",
		"scissorRectangle",
		"textureScale",
		"uniforms"
	]
}), Pn = Z({
	name: "BlackAndWhiteStage",
	props: ["gradations"],
	create: () => I.createBlackAndWhiteStage()
}), Fn = Z({
	name: "BrightnessStage",
	props: ["brightness"],
	create: () => I.createBrightnessStage()
}), In = Z({
	name: "LensFlareStage",
	props: [
		"dirtTexture",
		"starTexture",
		"intensity",
		"distortion",
		"ghostDispersal",
		"haloWidth",
		"earthRadius"
	],
	create: () => I.createLensFlareStage()
}), Ln = Z({
	name: "Fxaa",
	create: (e, t) => t.fxaa,
	props: []
}), Rn = Z({
	name: "NightVisionStage",
	props: [],
	create: () => I.createNightVisionStage()
}), zn = ["enabled", "selected"], Bn = [
	"inputPreviousStageTexture",
	"name",
	"stages",
	"uniforms"
], Vn = (e, t) => {
	let n = e, r = Array.isArray(n._stages) ? n._stages.slice() : [];
	if (n._stages = [], t) {
		t.remove(e);
		let n = t._stageNames;
		for (let e of r) {
			let t = e;
			n && typeof t.name == "string" && Reflect.deleteProperty(n, t.name), t._textureCache = void 0, t._index = void 0;
		}
	}
	e.isDestroyed() || e.destroy();
}, Hn = X({
	name: "PostProcessStageComposite",
	create(e, t) {
		if (!e.scene) return;
		let n = new Ee(t);
		return typeof t.enabled == "boolean" && (n.enabled = t.enabled), t.selected && (n.selected = t.selected), e.scene.postProcessStages.add(n), n;
	},
	destroy(e, t) {
		Vn(e, t.scene && !t.scene.isDestroyed() ? t.scene.postProcessStages : void 0);
	},
	cesiumProps: zn,
	cesiumReadonlyProps: Bn
}), Un = Z({
	name: "AmbientOcclusion",
	create: (e, t) => t.ambientOcclusion,
	props: [
		"ambientOcclusionOnly",
		"bias",
		"delta",
		"frustumLength",
		"intensity",
		"lengthCap",
		"sigma",
		"stepSize"
	],
	noMount: !0
}), Wn = Z({
	name: "Bloom",
	create: (e, t) => t.bloom,
	props: [
		"brightness",
		"contrast",
		"delta",
		"glowOnly",
		"sigma",
		"stepSize"
	],
	noMount: !0
}), Gn = Z({
	name: "BlurStage",
	props: [
		"delta",
		"sigma",
		"stepSize"
	],
	create: () => I.createBlurStage()
}), Kn = Z({
	name: "DepthOfFieldStage",
	props: [
		"delta",
		"focalDistance",
		"sigma",
		"stepSize"
	],
	create: () => I.createDepthOfFieldStage()
}), qn = Z({
	name: "EdgeDetectionStage",
	props: ["color", "length"],
	create: () => I.createEdgeDetectionStage()
}), Jn = Z({
	name: "SilhouetteStage",
	props: ["color", "length"],
	create: () => I.createSilhouetteStage()
}), Yn = X({
	name: "Primitive",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new De(t);
		if (t.onReady) {
			let r = () => {
				n.ready && (t.onReady?.(n), e.scene?.postRender.removeEventListener(r));
			};
			e.scene?.postRender.addEventListener(r);
		}
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: [
		"appearance",
		"cull",
		"debugShowBoundingVolume",
		"depthFailAppearance",
		"modelMatrix",
		"shadows",
		"show"
	],
	cesiumReadonlyProps: [
		"allowPicking",
		"asynchronous",
		"compressVertices",
		"geometryInstances",
		"interleave",
		"releaseGeometryInstances",
		"vertexCacheOptimize"
	],
	otherProps: ["onReady"],
	useCommonEvent: !0
}), Xn = X({
	name: "RectangleGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new Oe(t);
		return e.entity.rectangle = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.rectangle = void 0);
	},
	cesiumProps: [
		"classificationType",
		"coordinates",
		"height",
		"heightReference",
		"extrudedHeight",
		"extrudedHeightReference",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"rotation",
		"stRotation",
		"granularity",
		"shadows",
		"distanceDisplayCondition",
		"zIndex"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), Zn = /* @__PURE__ */ "backgroundColor.completeMorphOnUserInput.debugShowCommands.debugShowDepthFrustum.debugShowFramesPerSecond.debugShowFrustumPlanes.debugShowFrustums.eyeSeparation.farToNearRatio.focalLength.fog.gamma.globe.highDynamicRange.invertClassification.invertClassificationColor.light.logarithmicDepthBuffer.logarithmicDepthFarToNearRatio.maximumRenderTimeChange.minimumDisableDepthTestDistance.moon.morphTime.nearToFarDistance2D.pickTranslucentDepth.requestRenderMode.rethrowRenderErrors.shadowMap.skyAtmosphere.skyBox.specularEnvironmentMaps.sphericalHarmonicCoefficients.sun.sunBloom.terrainProvider.useDepthPicking.useWebVR.msaaSamples.splitPosition.verticalExaggeration.verticalExaggerationRelativeHeight.atmosphere".split("."), Qn = {
	onMorphComplete: "morphComplete",
	onMorphStart: "morphStart",
	onPostRender: "postRender",
	onPreRender: "preRender",
	onPreUpdate: "preUpdate",
	onPostUpdate: "postUpdate",
	onRenderError: "renderError",
	onTerrainProviderChange: "terrainProviderChanged"
}, $n = [
	"mode",
	"morphDuration",
	"enableEdgeVisibility"
], er = (e, t, n) => {
	switch (t) {
		case R.SCENE2D:
			e.morphTo2D(n);
			break;
		case R.COLUMBUS_VIEW:
			e.morphToColumbusView(n);
			break;
		case R.SCENE3D:
			e.morphTo3D(n);
			break;
	}
}, tr = X({
	name: "Scene",
	create(e, t) {
		return e.scene && (t.mode && er(e.scene, t.mode, t.morphDuration), t.enableEdgeVisibility !== void 0 && (e.scene._enableEdgeVisibility = t.enableEdgeVisibility)), e.scene;
	},
	update(e, t, n) {
		t.mode !== n.mode && t.mode && er(e, t.mode, t.morphDuration), t.enableEdgeVisibility !== n.enableEdgeVisibility && t.enableEdgeVisibility !== void 0 && (e._enableEdgeVisibility = t.enableEdgeVisibility);
	},
	cesiumProps: Zn,
	cesiumEventProps: Qn,
	otherProps: $n,
	setCesiumPropsAfterCreate: !0
}), nr = X({
	name: "ScreenSpaceCameraController",
	create: (e) => e.scene?.screenSpaceCameraController,
	cesiumProps: [
		"bounceAnimationTime",
		"enableCollisionDetection",
		"enableInputs",
		"enableLook",
		"enableRotate",
		"enableTilt",
		"enableTranslate",
		"enableZoom",
		"inertiaSpin",
		"inertiaTranslate",
		"inertiaZoom",
		"lookEventTypes",
		"maximumMovementRatio",
		"maximumZoomDistance",
		"minimumCollisionTerrainHeight",
		"minimumPickingTerrainHeight",
		"minimumTrackBallHeight",
		"minimumZoomDistance",
		"rotateEventTypes",
		"tiltEventTypes",
		"translateEventTypes",
		"zoomEventTypes",
		"minimumPickingTerrainDistanceWithInertia",
		"maximumTiltAngle",
		"zoomFactor"
	],
	setCesiumPropsAfterCreate: !0
}), rr = ({ action: e, modifier: t, type: n }) => {
	let r = U(), a = s(() => Array.isArray(t) ? [...t].sort((e, t) => e - t).join(",") : t ?? "", [t]);
	return i(() => {
		if (!(!r.screenSpaceEventHandler || r.screenSpaceEventHandler.isDestroyed())) {
			if (e) return r.screenSpaceEventHandler.setInputAction(e, n, t), () => {
				!r.screenSpaceEventHandler || r.screenSpaceEventHandler.isDestroyed() || r.screenSpaceEventHandler.removeInputAction(n, t);
			};
			r.screenSpaceEventHandler.removeInputAction(n, t);
		}
	}, [
		e,
		r.screenSpaceEventHandler,
		a,
		n
	]), null;
}, ir = X({
	name: "ScreenSpaceEventHandler",
	create(e, t) {
		return t.useDefault ? e.cesiumWidget?.screenSpaceEventHandler : e.scene ? new ke(e.scene.canvas) : void 0;
	},
	destroy(e) {
		e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { screenSpaceEventHandler: e };
	}
}), ar = X({
	name: "ShadowMap",
	create: (e) => e.scene?.shadowMap,
	cesiumProps: [
		"darkness",
		"fadingEnabled",
		"maximumDistance",
		"enabled",
		"normalOffset",
		"softShadows",
		"size"
	]
}), or = X({
	name: "SkyAtmosphere",
	create: (e) => e.scene?.skyAtmosphere,
	cesiumProps: [
		"brightnessShift",
		"hueShift",
		"saturationShift",
		"show",
		"perFragmentAtmosphere",
		"atmosphereLightIntensity",
		"atmosphereRayleighCoefficient",
		"atmosphereMieCoefficient",
		"atmosphereRayleighScaleHeight",
		"atmosphereMieScaleHeight",
		"atmosphereMieAnisotropy"
	],
	setCesiumPropsAfterCreate: !0
}), sr = X({
	name: "SkyBox",
	create: (e) => e.scene?.skyBox,
	cesiumProps: ["sources", "show"],
	setCesiumPropsAfterCreate: !0
}), cr = X({
	name: "Sun",
	create(e) {
		if (!e.scene) return;
		let t = new Ae();
		return e.scene.sun = t, t;
	},
	destroy(e, t) {
		t.scene && !t.scene.isDestroyed() && (t.scene.sun = new Ae());
	},
	cesiumProps: ["glowFactor", "show"],
	setCesiumPropsAfterCreate: !0
}), lr = X({
	name: "TimeDynamicPointCloud",
	create(e, t) {
		if (!e.cesiumWidget || !e.primitiveCollection || !e.cesiumWidget?.clock) return;
		let n = new je({
			...t,
			clock: t.clock ?? e.cesiumWidget.clock
		});
		if (t.onReady) {
			let e = () => {
				t.onReady?.(n), n.frameChanged.removeEventListener(e);
			};
			n.frameChanged.addEventListener(e);
		}
		return e.primitiveCollection.add(n), [n, { userClippingPlanes: !!t.clippingPlanes }];
	},
	destroy(e, t, n, r) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || (r?.userClippingPlanes && (e._clippingPlanes = void 0), e.destroy());
	},
	cesiumProps: [
		"clippingPlanes",
		"maximumMemoryUsage",
		"modelMatrix",
		"shadows",
		"show",
		"style",
		"intervals"
	],
	cesiumReadonlyProps: ["clock", "shading"],
	cesiumEventProps: { onFrameChange: "frameChanged" },
	otherProps: ["onReady"],
	useCommonEvent: !0
}), ur = X({
	name: "Viewer",
	async create(e, { baseLayer: t, terrainProvider: n, ...r }, i) {
		if (!i) return;
		let a;
		a = q(n) ? await n : n;
		let o = new Me(i, {
			...r,
			terrainProvider: a,
			baseLayer: t ?? void 0
		});
		if (o) return t === !1 && o.imageryLayers.removeAll(), o && r.extend && (Array.isArray(r.extend) ? r.extend.forEach((e) => {
			o.extend(e, {});
		}) : o.extend(r.extend, {})), [o, new Ue(o.scene)];
	},
	destroy(e, t, n, r) {
		r && !r.isDestroyed() && r.destroy(), e.isDestroyed() || e.destroy();
	},
	provide(e, t, n, r) {
		return {
			viewer: e,
			cesiumWidget: e.cesiumWidget,
			dataSourceCollection: e.dataSources,
			entityCollection: e.entities,
			scene: e.scene,
			camera: e.scene.camera,
			imageryLayerCollection: e.scene.globe.imageryLayers,
			primitiveCollection: e.scene.primitives,
			globe: e.scene.globe,
			__$internal: {
				onUpdate: n?.onUpdate,
				imageryLayerWaitingList: []
			},
			[J]: r
		};
	},
	containerProps: ({ id: e, className: t, style: n, full: r, containerProps: i }) => ({
		className: t,
		id: e,
		style: {
			...r ? {
				position: "absolute",
				bottom: "0",
				left: "0",
				right: "0",
				top: "0"
			} : {},
			...n
		},
		...i
	}),
	cesiumProps: [
		"terrainShadows",
		"clockTrackedDataSource",
		"targetFrameRate",
		"useDefaultRenderLoop",
		"resolutionScale",
		"allowDataSourcesToSuspendAnimation",
		"trackedEntity",
		"selectedEntity",
		"shadows",
		"useBrowserRecommendedResolution",
		"creditDisplay"
	],
	cesiumReadonlyProps: /* @__PURE__ */ "baseLayer.animation.baseLayerPicker.fullscreenButton.vrButton.geocoder.homeButton.infoBox.sceneModePicker.selectionIndicator.timeline.navigationHelpButton.navigationInstructionsInitiallyVisible.scene3DOnly.shouldAnimate.clockViewModel.selectedImageryProviderViewModel.imageryProviderViewModels.selectedTerrainProviderViewModel.terrainProviderViewModels.ellipsoid.skyBox.skyAtmosphere.fullscreenElement.showRenderLoopErrors.automaticallyTrackDataSourceClocks.contextOptions.sceneMode.mapProjection.globe.orderIndependentTranslucency.creditContainer.creditViewport.dataSources.mapMode2D.projectionPicker.blurActiveElementOnCanvasFocus.requestRenderMode.maximumRenderTimeChange.depthPlaneEllipsoidOffset.msaaSamples.terrain".split("."),
	cesiumEventProps: {
		onSelectedEntityChange: "selectedEntityChanged",
		onTrackedEntityChange: "trackedEntityChanged"
	},
	otherProps: [
		"className",
		"id",
		"style",
		"full",
		"containerProps",
		"extend",
		"terrainProvider"
	],
	renderContainer: !0,
	useCommonEvent: !0,
	useRootEvent: !0
}), dr = X({
	name: "WebMapTileServiceImageryProvider",
	create(e, t) {
		let { onReady: n, children: r, ...i } = t, a = new Pe(i);
		return n && n(a), a;
	}
}), fr = X({
	name: "WallGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new Ne(t);
		return e.entity.wall = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.wall = void 0);
	},
	cesiumProps: [
		"positions",
		"maximumHeights",
		"minimumHeights",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"granularity",
		"shadows",
		"distanceDisplayCondition"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
});
//#endregion
export { Un as AmbientOcclusion, Qe as Billboard, $e as BillboardCollection, et as BillboardGraphics, Pn as BlackAndWhiteStage, Wn as Bloom, Gn as BlurStage, tt as BoxGraphics, Fn as BrightnessStage, nt as BufferPoint, rt as BufferPointCollection, it as BufferPolygon, at as BufferPolygonCollection, ot as BufferPolyline, st as BufferPolylineCollection, ct as Camera, lt as CameraFlyHome, ut as CameraFlyTo, ft as CameraFlyToBoundingSphere, dt as CameraLookAt, xt as Cesium3DTilesTerrainProvider, yt as Cesium3DTileset, bt as Cesium3DTilesetGraphics, H as CesiumContext, St as CesiumWidget, Ct as ClassificationPrimitive, wt as Clock, Tt as CloudCollection, Re as Consumer, Et as CorridorGraphics, Dt as CubeMapPanorama, Ot as CumulusCloud, kt as CustomDataSource, At as CylinderGraphics, Lt as CzmlDataSource, Kn as DepthOfFieldStage, qn as EdgeDetectionStage, Rt as EllipseGraphics, zt as EllipsoidGraphics, Bt as Entity, Ht as EntityDescription, Vt as EquirectangularPanorama, Ue as EventManager, Ut as Fog, Ln as Fxaa, Xt as GeoJsonDataSource, Zt as GeoJsonPrimitive, Qt as Globe, $t as Google2DImageryProvider, en as GooglePhotorealistic3DTileset, tn as GroundPolylinePrimitive, nn as GroundPrimitive, rn as GroundPrimitiveCollection, an as ImageryLayer, on as ImageryLayerCollection, pn as KmlDataSource, mn as Label, hn as LabelCollection, gn as LabelGraphics, In as LensFlareStage, bn as MVTDataProvider, _n as Model, vn as ModelGraphics, yn as Moon, Rn as NightVisionStage, Sn as ParticleSystem, Cn as PathGraphics, wn as PlaneGraphics, Tn as PointGraphics, En as PointPrimitive, Dn as PointPrimitiveCollection, On as PolygonGraphics, kn as Polyline, An as PolylineCollection, jn as PolylineGraphics, Mn as PolylineVolumeGraphics, Nn as PostProcessStage, Hn as PostProcessStageComposite, Yn as Primitive, Le as Provider, Xn as RectangleGraphics, tr as Scene, nr as ScreenSpaceCameraController, rr as ScreenSpaceEvent, ir as ScreenSpaceEventHandler, ar as ShadowMap, Jn as SilhouetteStage, or as SkyAtmosphere, sr as SkyBox, cr as Sun, lr as TimeDynamicPointCloud, ur as Viewer, fr as WallGraphics, dr as WebMapTileServiceImageryProvider, Ye as clearCesiumResource, W as createCameraOperation, X as createCesiumComponent, Z as createPostProcessStage, G as entries, J as eventManagerContextKey, Y as eventNames, xn as getTileset, K as includes, Ve as isDestroyable, He as isDestroyed, q as isPromise, ze as pick, Xe as resourceUrl, Be as shallowEquals, U as useCesium, We as useCesiumComponent, Je as useCesiumResource, $ as useSuspendedResource, Ze as withCesium };
